### PR TITLE
feat: stopwatch

### PR DESCRIPTION
**Pull request by caarlos0**
follow up of #67, a simple stopwatch component

### CODE REVIEW

File: stopwatch/stopwatch.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+// NewWithInterval creates a new timer with the given timeout and tick interval.
+func NewWithInterval(interval time.Duration) Model {
+	return Model{
+		TickEvery: interval,
+	}
+}
+
+// New creates a new timer with the given timeout and default 1s interval.
+func New(timeout time.Duration) Model {
```
Comment (meowgorithm) [inline stopwatch/stopwatch.go:32]: We should drop the arg here (and change the comment) since it defaults to 1s

File: stopwatch/stopwatch.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+// TickMsg is a message that is sent on every timer tick.
+type TickMsg struct{}
+
+// StopMsg is a message that can be send to stop the watch.
+type StopMsg struct{}
+
+// Model of the timer component.
+type Model struct {
```
Comment (meowgorithm) [inline stopwatch/stopwatch.go:17]: Should we add a method for returning the time elapsed since the stopwatch started?

**Comment by meowgorithm** [comment]
So in cases like this, `StopMsg` won't stop the timer:

```go
package main

import (
	"fmt"
	"os"
	"time"

	"github.com/charmbracelet/bubbles/stopwatch"
	tea "github.com/charmbracelet/bubbletea"
)

type model struct {
	stopwatch stopwatch.Model
}

func (m model) Init() tea.Cmd {
	return m.stopwatch.Init()
}

func (m model) View() string {
	return "Elapsed: " + m.stopwatch.View()
}

func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch msg.String() {
		case "ctrl+c", "q":
			return m, tea.Quit
		case "s":
			return m, func() tea.Msg {
				return stopwatch.StopMsg{}
			}
		}
	}
	var cmd tea.Cmd
	m.stopwatch, cmd = m.stopwatch.Update(msg)
	return m, cmd
}

func main() {
	m := model{
		stopwatch: stopwatch.NewWithInternval(time.Millisecond),
	}

	if err := tea.NewProgram(m).Start(); err != nil {
		fmt.Println("Oh no, it didn't work:", err)
		os.Exit(1)
	}
}
```

Though technically we can stop the timer without `StopMsg`:

```go
package main

import (
	"fmt"
	"os"
	"time"

	"github.com/charmbracelet/bubbles/stopwatch"
	tea "github.com/charmbracelet/bubbletea"
)

type model struct {
	stopwatch stopwatch.Model
	runTimer  bool
}

func (m model) Init() tea.Cmd {
	return m.stopwatch.Init()
}

func (m model) View() string {
	return "Elapsed: " + m.stopwatch.View()
}

func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch msg.String() {
		case "ctrl+c", "q":
			return m, tea.Quit
		case "s":
			m.runTimer = !m.runTimer
			if m.runTimer {
				return m, m.stopwatch.Init()
			}
		}
	}

	var cmd tea.Cmd
	if m.runTimer {
		m.stopwatch, cmd = m.stopwatch.Update(msg)
	}
	return m, cmd
}

func main() {
	m := model{
		stopwatch: stopwatch.NewWithInterval(time.Millisecond),
		runTimer:  true,
	}

	if err := tea.NewProgram(m).Start(); err != nil {
		fmt.Println("Oh no, it didn't work:", err)
		os.Exit(1)
	}
}
```

…so I wonder if we need it at all? If we do, maybe `Model.StopMsg()` is a better choice since we don't need to send a command through the runtime?

**Comment by caarlos0** [comment]
Make it a bit more feature-full:

- start
- stop
- reset
- get elapsed time
- get if its running or not

usage now looks like this:

```go
package main

import (
	"fmt"
	"os"
	"time"

	"github.com/charmbracelet/bubbles/stopwatch"
	tea "github.com/charmbracelet/bubbletea"
)

type model struct {
	stopwatch stopwatch.Model
}

func (m model) Init() tea.Cmd {
	return m.stopwatch.Init()
}

func (m model) View() string {
	return "Elapsed: " + m.stopwatch.View()
}

func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch msg.String() {
		case "ctrl+c", "q":
			return m, tea.Quit
		case "r":
			return m, m.stopwatch.Reset()
		case "s":
			if m.stopwatch.Running() {
				return m, m.stopwatch.Stop()
			}
			return m, m.stopwatch.Start()
		}
	}
	var cmd tea.Cmd
	m.stopwatch, cmd = m.stopwatch.Update(msg)
	return m, cmd
}

func main() {
	m := model{
		stopwatch: stopwatch.NewWithInterval(time.Second),
	}

	if err := tea.NewProgram(m).Start(); err != nil {
		fmt.Println("Oh no, it didn't work:", err)
		os.Exit(1)
	}
}
```

thoughts?

File: stopwatch/stopwatch.go (hunk 2 above)
Comment (caarlos0) [reply stopwatch/stopwatch.go:20]: done !

**Comment by meowgorithm** [comment]
Looks great! Yeah, we might as well add a `Toggle` function, then I think this is good to go.

**Pull request merged**